# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/ArrayBufferViewObject.cpp
# Commit: 2865fe682139
# Full Hash: 2865fe682139450ad7574810366f980d771412c2
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2023-09-12 04:12:49
# Regressor Bug: 1690111
# File Overlap Count: 1
# Description:
#   Bug 1690111 - Implement the ability to pin lengths (prevent changes to length, eg via detaching) for ArrayBuffers and their views. r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D131301
# ==============================================================================

diff -r 9dcd2416f8a5 -r 2865fe682139 js/src/vm/ArrayBufferViewObject.cpp
--- a/js/src/vm/ArrayBufferViewObject.cpp	Mon Sep 11 12:52:18 2023 +0000
+++ b/js/src/vm/ArrayBufferViewObject.cpp	Mon Sep 11 12:52:19 2023 +0000
@@ -56,6 +56,7 @@
 void ArrayBufferViewObject::notifyBufferDetached() {
   MOZ_ASSERT(!isSharedMemory());
   MOZ_ASSERT(hasBuffer());
+  MOZ_ASSERT(!bufferUnshared()->isLengthPinned());
 
   setFixedSlot(LENGTH_SLOT, PrivateValue(size_t(0)));
   setFixedSlot(BYTEOFFSET_SLOT, PrivateValue(size_t(0)));
@@ -98,7 +99,12 @@
 
   initFixedSlot(BYTEOFFSET_SLOT, PrivateValue(byteOffset));
   initFixedSlot(LENGTH_SLOT, PrivateValue(length));
-  initFixedSlot(BUFFER_SLOT, ObjectOrNullValue(buffer));
+  if (buffer) {
+    initFixedSlot(BUFFER_SLOT, ObjectValue(*buffer));
+  } else {
+    MOZ_ASSERT(!isSharedMemory());
+    initFixedSlot(BUFFER_SLOT, JS::FalseValue());
+  }
 
   if (buffer) {
     SharedMem<uint8_t*> ptr = buffer->dataPointerEither();
@@ -317,3 +323,18 @@
   return false;
 #endif
 }
+
+JS_PUBLIC_API bool JS::PinArrayBufferOrViewLength(JSObject* obj, bool pin) {
+  ArrayBufferObjectMaybeShared* buffer =
+      obj->maybeUnwrapIf<ArrayBufferObjectMaybeShared>();
+  if (buffer) {
+    return buffer->pinLength(pin);
+  }
+
+  ArrayBufferViewObject* view = obj->maybeUnwrapAs<ArrayBufferViewObject>();
+  if (view) {
+    return view->pinLength(pin);
+  }
+
+  return false;
+}